packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,201 @@
+    );
+    const expectedNodeKeys = Object.keys(expectedNode).filter(
+      key => !astIgnoreKeys.includes(key),
+    );
+    if (actualNodeKeys.length !== expectedNodeKeys.length) {
+      return false;
+    }
+    if (
+      actualNodeKeys.some(
+        actualNodeKey =>
+          !Object.prototype.hasOwnProperty.call(expectedNode, actualNodeKey),
+      )
+    ) {
+      return false;
+    }
+    if (
+      actualNodeKeys.some(
+        actualNodeKey =>
+          !isSameAstNode(
+            actualNode[actualNodeKey as keyof typeof actualNode],
+            expectedNode[actualNodeKey as keyof typeof expectedNode],
+          ),
+      )
+    ) {
+      return false;
+    }
+    return true;
+  }
+  return false;
+};
+
+export default util.createRule<Options, MessageIds>({
+  name: 'no-duplicate-type-constituents',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Disallow duplicate constituents of union or intersection types',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    fixable: 'code',
+    messages: {
+      duplicate:
+        '{{type}} type member {{duplicated}} is duplicated with {{previous}}.',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreIntersections: {
+            type: 'boolean',
+          },
+          ignoreUnions: {
+            type: 'boolean',
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      ignoreIntersections: false,
+      ignoreUnions: false,
+    },
+  ],
+  create(context, [{ ignoreIntersections, ignoreUnions }]) {
+    const sourceCode = context.getSourceCode();
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    function checkDuplicate(
+      node: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
+    ): void {
+      const duplicateConstituents: {
+        duplicated: TSESTree.TypeNode;
+        duplicatePrevious: TSESTree.TypeNode;
+      }[] = [];
+
+      node.types.reduce<
+        {
+          node: TSESTree.TypeNode;
+          type: Type;
+        }[]
+      >((uniqConstituents, constituentNode) => {
+        const type = checker.getTypeAtLocation(
+          parserServices.esTreeNodeToTSNodeMap.get(constituentNode),
+        );
+        const duplicatePreviousConstituent = uniqConstituents.find(
+          ele => isSameAstNode(ele.node, constituentNode) || ele.type === type,
+        );
+        if (duplicatePreviousConstituent) {
+          duplicateConstituents.push({
+            duplicated: constituentNode,
+            duplicatePrevious: duplicatePreviousConstituent.node,
+          });
+          return uniqConstituents;
+        }
+        return [...uniqConstituents, { node: constituentNode, type }];
+      }, []);
+
+      const fix: TSESLint.ReportFixFunction = fixer => {
+        return duplicateConstituents
+          .map(duplicateConstituent => {
```

Comment:
Nit: this way it doesn't need to `.flat()` later ✨ 

```suggestion
          .flatMap(duplicateConstituent => {
```

Reply:
You're absolutely right! (Why didn't I think of that? .... 🤦‍♂️)